// redux-remember/dist/cjs/persist.cjs
"use strict";

exports.__esModule = true;
exports.saveAllKeyed = exports.saveAll = exports.persist = void 0;
var _errors = require("./errors.cjs");
var _isDeepEqual = _interopRequireDefault(require("./is-deep-equal.cjs"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
var saveAll = exports.saveAll = function saveAll(state, oldState, _ref) {
  var prefix = _ref.prefix,
    driver = _ref.driver,
    serialize = _ref.serialize;
  if (!(0, _isDeepEqual["default"])(state, oldState)) {
    return driver.setItem(prefix + "rootState", serialize(state, 'rootState'));
  }
};
var saveAllKeyed = exports.saveAllKeyed = function saveAllKeyed(state, oldState, _ref2) {
  var prefix = _ref2.prefix,
    driver = _ref2.driver,
    serialize = _ref2.serialize;
  return Promise.all(Object.keys(state).map(function (key) {
    if ((0, _isDeepEqual["default"])(state[key], oldState[key])) {
      return Promise.resolve();
    }
    return driver.setItem("" + prefix + key, serialize(state[key], key));
  }));
};
var persist = exports.persist = /*#__PURE__*/function () {
  var _ref4 = _asyncToGenerator(function* (state, oldState, _ref3) {
    if (state === void 0) {
      state = {};
    }
    if (oldState === void 0) {
      oldState = {};
    }
    var prefix = _ref3.prefix,
      driver = _ref3.driver,
      persistWholeStore = _ref3.persistWholeStore,
      serialize = _ref3.serialize,
      errorHandler = _ref3.errorHandler;
    try {
      var save = persistWholeStore ? saveAll : saveAllKeyed;
      yield save(state, oldState, {
        prefix: prefix,
        driver: driver,
        serialize: serialize
      });
    } catch (err) {
      errorHandler(new _errors.PersistError(err));
    }
  });
  return function persist(_x, _x2, _x3) {
    return _ref4.apply(this, arguments);
  };
}();
